polkadot/node/core/prospective-parachains/src/lib.rs

```diff
@@ -280,119 +282,113 @@ async fn handle_active_leaves_update<Context>(
+			let prev_fragment_chain =
+				prev_fragment_chains.and_then(|chains| chains.fragment_chains.get(&para));
+
+			if let Some(prev_fragment_chain) = prev_fragment_chain {
+				// Add old candidates to the new storage only after we added the pending
+				// availability candidates. The pending candidates have higher priority and can
+				// conflict with the old candidates.
+				for candidate in prev_fragment_chain.as_candidate_storage().into_candidates() {
+					// If they used to be pending availability, don't add them. This is fine
+					// because:
+					// - if they still are pending availability, they have already been added to the
+					//   new storage.
+					// - if they were included, no point in keeping them.
+					if prev_fragment_chain
+						.scope()
+						.get_pending_availability(&candidate.hash())
+						.is_none()
+					{
+						// We need to swallow any potential errors here, as they can happen under
+						// normal operation, with candidates becoming out of scope for example.
+						let _ = new_storage.add_candidate_entry(candidate);
+					}
+				}
+			}
 
 			gum::trace!(
 				target: LOG_TARGET,
 				relay_parent = ?hash,
 				min_relay_parent = scope.earliest_relay_parent().number,
 				para_id = ?para,
+				ancestors = ?ancestry,
 				"Creating fragment chain"
 			);
 
-			let chain = FragmentChain::populate(scope, &*candidate_storage);
+			// Finally, populate the fragment chain.
+			let chain = FragmentChain::populate(scope, new_storage);
 
 			gum::trace!(
 				target: LOG_TARGET,
 				relay_parent = ?hash,
 				para_id = ?para,
-				"Populated fragment chain with {} candidates",
-				chain.len()
+				"Populated fragment chain with {} candidates: {:?}",
+				chain.best_chain_len(),
+				chain.best_chain_vec()
+			);
+
+			gum::trace!(
+				target: LOG_TARGET,
+				relay_parent = ?hash,
+				para_id = ?para,
+				"Potential candidate storage for para: {:?}",
+				chain.unconnected().map(|candidate| candidate.hash()).collect::<Vec<_>>()
 			);
 
 			fragment_chains.insert(para, chain);
 		}
 
-		view.active_leaves.insert(
-			hash,
-			RelayBlockViewData { fragment_chains, pending_availability: all_pending_availability },
-		);
+		view.active_leaves.insert(hash, RelayBlockViewData { fragment_chains });
 	}
 
-	if !update.deactivated.is_empty() {
-		// This has potential to be a hotspot.
-		prune_view_candidate_storage(view, metrics);
-	}
-
-	Ok(())
-}
-
-fn prune_view_candidate_storage(view: &mut View, metrics: &Metrics) {
-	let _timer = metrics.time_prune_view_candidate_storage();
-
-	let active_leaves = &view.active_leaves;
-	let mut live_candidates = HashSet::new();
-	let mut live_paras = HashSet::new();
-	for sub_view in active_leaves.values() {
-		live_candidates.extend(sub_view.pending_availability.iter().cloned());
-
-		for (para_id, fragment_chain) in &sub_view.fragment_chains {
-			live_candidates.extend(fragment_chain.to_vec());
-			live_paras.insert(*para_id);
-		}
+	for deactivated in &update.deactivated {
```

Comment:
There's a bug here. We should keep deactivated leaves until they go out of scope (like backing_implicit_view) does, so that a subsequent fork may get the same candidates from the parent as the first fork did.

Reply:
fixed